### DISCUSSION TITLE
Spatial calibration - wire intersections not visible

### DISCUSSION THREAD

**Discussion by NesBarn** [PlusToolkit/PlusLib#897] [Q&A]
Hello,

By changing the ultrasound system (terason), I encountered a problem of visibility of the phantom wire intersections.
I followed the recommandations by adapting the gain to make the ultrasound image mostly black. I tried also to play with the spacing but still no dots (even red) are visible... Here is a screenshot of the ultrasound image I have. If you have any idea/recommandation, I would be grateful!

Thank you
![image](https://user-images.githubusercontent.com/97302106/151774805-59d6d468-1831-40f4-bb54-4e7f13cf09c7.png)

**Comment by Sunderlandkyl** [comment]
I'm not familiar with the process used for wire detection. @lassoan or @adamrankin do you have any suggestions?

**Reply by adamrankin** [reply]
Can you take a recording in mha or nrrd format and upload it for us to play with? I can try to find settings that will segment, but it may be that you have to play more with US settings.

**Reply by NesBarn** [reply]
Hi Adam,
What do you like to have exactly: a recording of an ultrasound aquistion of the phantom, like a loop? or just the ultrasound image recorded in nrrd format? I have played with the gain and focus to try to have the image almost black and the wires bright. 
To answer to Andreas, I freezed the ultrasound image, used rulers to get the distance and put it on the tab but still no red dots over the wires.

**Reply by NesBarn** [reply]
I tried to send an image in nrrd format but I can't, this file type isn't supported here..

**Reply by lassoan** [reply]
You cannot upload file attachments, you need to upload to dropbox/onedrive/gdrive/etc. and post the link here.

**Reply by NesBarn** [reply]
Here it is: https://we.tl/t-SX0jaS2YQk
Do you want also the image I sent below with the rulers?

**Comment by lassoan** [comment]
The image above should be good enough for dots to be detected. Most likely the issue is that the approximate spacing is incorrect.

> Check if initial spacing is correct. For giving a hint on image spacing to fCal, select the checkbox next to spacing. It will give you two rulers, a green (G) and a blue (B). You can manually enter two known distances that appear on the image, one horizontal and one vertical. Then set the visual rulers on the image to represent these distances. You can do this by dragging the end points of the ruler.

(source: http://perk-software.cs.queensu.ca/plus/doc/nightly/user/ApplicationfCal.html)

**Reply by NesBarn** [reply]
Yes, it is exactly what I did but it doesn't work... We can see in the image below the segments green and blue over the rulers and still there is no dot over the bright points .
![image](https://user-images.githubusercontent.com/97302106/152559579-458be474-b7e9-4093-95fb-f605016a1b71.png)

**Reply by lassoan** [reply]
The green and blue line are approximately the same length in the image (about 15% difference between them), while you set 20mm and 30mm for physical length. This is such a huge difference that probably does not fit into the default approximate spacing tolerance.

Check if the horizontal wire distance is really 30mm and the vertical distance is really 20mm. It seems more likely that the vertical distance is actually 30mm. If the physical distance is correct but on the image the vertical distance appears just so much larger then you may need to adjust the speed of sound in the ultrasound device or change the water temperature or use a different medium that better matches the speed of sound that is set in your ultrasound.

If you want to calibrate an image with an extremely anisotropic spacing (it is not a good idea, but if for some reason you must do it) then you can adjust the spacing tolerance value in the Plus Device Set configuration XML file.

**Reply by NesBarn** [reply]
![image](https://user-images.githubusercontent.com/97302106/152564671-3c3a819d-50d2-4f56-9185-46de521f7827.png)
Unfortunatelyn the distances are correct as we can see in this image 
Regarding the tolerances, I tried to change the collinear max distance, many red dots appeared but no one on the wire intersections

**Reply by lassoan** [reply]
I see that the vertical distance is 20mm according to the depth marker. However, there is no horizontal ruler. Please use the Caliper tool on the ultrasound system and measure the wire distances.

Please also post a screenshot when you see many candidate points. If the wire intersections are detected as candidates then that confirms that point detection works well (no need to change Morphology parameters or image acquisition settings).

If that works then you may increase all distance tolerances (`Line pair distance`, `Max line shift`) to let the pattern match succeed. Relaxing the tolerance may cause matching of incorrect candidates, so if you find that this happens then you probably need to fix the speed of sound or the specified distances.

**Reply by NesBarn** [reply]
Hello @lassoan 
Indeed because no horizontal ruler is present, I used the caliper tool as we can see on the previous image on the left (surrounded by the red circle). Anyway, I took new screenshots to show you what are the candidates that appear. I tried to increase all distance tolerances but still no accurate candidate appears...
![image](https://user-images.githubusercontent.com/97302106/152815352-bca2ad96-0c51-4e80-8d0a-835b9e6e9464.png)
![image](https://user-images.githubusercontent.com/97302106/152815558-04210297-6aea-4138-a911-ae8fb41b5957.png)

**Reply by lassoan** [reply]
There are two independent issues:

1. Highly anisotropic spacing in the acquired image

x spacing = 30mm / 425px = 0.0706
y spacing = 20mm / 369px = 0.0542

This is a 30% difference, which is significant. Especially because this is not related to difference in actual vs. assumed speed of sound, because these are the ultrasound system's own distance measurements. I have never seen an ultrasound system where the spacing along x and y axis did not match in images that are displayed to the user (and that's why only a vertical scale bar is sufficient). Do you see that the image that you see in fCal is distorted (vertically stretched out) compared to what you see on the ultrasound cart? If yes, then something may be wrong with the framegrabber hardware or configuration.

Plus can calibrate anisotropic images, as images are often slightly anisotropic due to speed of sound difference, but that is usually smaller (less than 10%). If there is a distortion of the image then it should be fixed also to ensure consistency between the images displayed on the cart and in the research application.

2. Wrong candidate points are detected

There is nothing that resembles to bright spots where the point candidates show up. The candidate point detection algorithm does not seem to work at all for your images.

Since it still works for the image in our test suite, most likely there is something about your data that is unexpected. Maybe the issue is that the point detection algorithm in fCal requires grayscale images. Could you try acquiring grayscale images and see if it fixes the issue?
